apiserver/facades/client/cloud/cloud_test.go

```diff
@@ -1376,267 +1865,10 @@ func (s *cloudSuite) TestCredentialContentsAllNoSecrets(c *gc.C) {
-}
-
-func (st *mockBackend) RemoveCloudAccess(cloud string, user names.UserTag) error {
-	st.MethodCall(st, "RemoveCloudAccess", cloud, user)
-	st.cloudAccess = permission.NoAccess
-	return nil
-}
-
-func (st *mockBackend) RemoveModelsCredential(tag names.CloudCredentialTag) error {
-	st.MethodCall(st, "RemoveModelsCredential", tag)
-	return st.NextErr()
-}
-
-type mockUser struct {
-	name string
-}
-
-func (m *mockUser) DisplayName() string {
-	return "display-" + m.name
-}
-
-type mockModel struct {
-	uuid               string
-	cloud              string
-	cloudRegion        string
-	cloudValue         cloud.Cloud
-	cloudCredentialTag names.CloudCredentialTag
-	cfg                *config.Config
-}
-
-func (m *mockModel) CloudName() string {
-	return m.cloud
-}
-
-func (m *mockModel) CloudRegion() string {
-	return m.cloudRegion
-}
-
-func (m *mockModel) CloudCredentialTag() (names.CloudCredentialTag, bool) {
-	return m.cloudCredentialTag, true
-}
-
-func (m *mockModel) Cloud() (cloud.Cloud, error) {
-	return m.cloudValue, nil
-}
-
-func (m *mockModel) CloudCredential() (state.Credential, bool, error) {
-	return state.Credential{}, false, nil
-}
-
-func (m *mockModel) ValidateCloudCredential(tag names.CloudCredentialTag, credential cloud.Credential) error {
-	return nil
-}
-
-func (m *mockModel) Config() (*config.Config, error) {
-	return m.cfg, nil
-}
-
-func (m *mockModel) Type() state.ModelType {
-	return state.ModelType("")
-}
-
-func (m *mockModel) UUID() string {
-	return m.uuid
-}
-
-type mockStatePool struct {
-	getF func(modelUUID string) (credentialcommon.PersistentBackend, context.ProviderCallContext, error)
-}
-
-func (m *mockStatePool) GetModelCallContext(modelUUID string) (credentialcommon.PersistentBackend, context.ProviderCallContext, error) {
-	return m.getF(modelUUID)
-}
-
-func (m *mockStatePool) SystemState() (*state.State, error) {
-	return nil, nil
-}
-
-type mockPooledModel struct {
-	release bool
-	model   *mockModelBackend
-}
-
-func (m *mockPooledModel) Model() credentialcommon.PersistentBackend {
-	return m.model
-}
-
-func (m *mockPooledModel) Release() bool {
-	return m.release
-}
-
-type mockModelBackend struct {
-	uuid  string
-	model *mockModel
-}
-
-func (m *mockModelBackend) Model() (credentialcommon.Model, error) {
-	return m.model, nil
-}
-
-func (m *mockModelBackend) Cloud(name string) (cloud.Cloud, error) {
-	return m.model.cloudValue, nil
-}
-
-func (m *mockModelBackend) AllMachines() ([]credentialcommon.Machine, error) {
-	return nil, nil
-}
-
-func (m *mockModelBackend) CloudCredential(tag names.CloudCredentialTag) (state.Credential, error) {
-	return state.Credential{}, nil
-}
-
-func (m *mockModelBackend) ControllerConfig() (credentialcommon.ControllerConfig, error) {
-	return nil, errors.NotImplementedf("ControllerConfig")
+func cloudCredentialTag(name, owner, cloudName string, attrs map[string]string) state.Credential {
+	cred := statetesting.CloudCredential(jujucloud.EmptyAuthType, attrs)
```

Comment:
`jujucloud.EmptyAuthType` isn't always what you want here. It is for credentialOne, but credentialTwo uses `jujucloud.UserPassAuthType`

Reply:
Refactored, now passing the permission as an function argument.

Reply:
Refactored once more the `cloudCredentialTag` now accepts 2 arguments, a struct and  the `gc.checker` type for internal assertion of errors and returns both a `state.Credential` and a `names.CloudCredentialTag`